### ISSUE TITLE
vscode like complete experience

### ISSUE THREAD

**Issue by sandangel** [lifepillar/vim-mucomplete#131]
I come here after reading [this post on reddit](https://www.reddit.com/r/vim/comments/9xnynn/has_anyone_reproduced_vscodes_autocomplete/)

Could you please help me add an example config for this in the wiki or readme? This is what I got so far but it does not work.

```sh
set completeopt+=menuone,noinsert
set shortmess+=Icm    " Shut off completion messages

" vim-mucomplete
let g:mucomplete#enable_auto_at_startup = 1
let g:mucomplete#no_mappings = 1
let g:mucomplete#no_popup_mappings = 1
let g:mucomplete#always_use_completeopt = 1
let g:mucomplete#buffer_relative_paths = 1
let g:mucomplete#completion_delay = 300

" imap <unique> <tab> <plug>(MUcompleteFwd)
" imap <unique> <s-tab> <plug>(MUcompleteBwd)
" inoremap <silent> <plug>(MUcompleteFwdKey) <down>
" imap <down> <plug>(MUcompleteCycFwd)
" inoremap <silent> <plug>(MUcompleteBwdKey) <up>
" imap <up> <plug>(MUcompleteCycBwd)
imap <unique> <c-e> <plug>(MUcompletePopupCancel)
imap <unique> <tab> <plug>(MUcompletePopupAccept)

```

I had setup LanguageClient but hit `.` in `.ts` file does not trigger auto completion popup too.

```sh

let g:LanguageClient_serverCommands = {}

nnoremap <silent> gh :call LanguageClient#textDocument_hover()<CR>
nnoremap <silent> gd :call LanguageClient#textDocument_definition()<CR>
nnoremap <silent> <F2> :call LanguageClient#textDocument_rename()<CR>

" yarn global add vscode-css-languageserver-bin
" https://github.com/vscode-langservers/vscode-css-languageserver-bin
if executable('css-languageserver')
  let g:LanguageClient_serverCommands = extend(g:LanguageClient_serverCommands, {
  \ 'css': ['css-languageserver', '--stdio' ],
  \ 'scss': ['css-languageserver', '--stdio' ],
  \ 'sass': ['css-languageserver', '--stdio' ],
  \ 'less': ['css-languageserver', '--stdio' ],
  \ })
endif

" yarn global add javascript-typescript-langserver
if executable('javascript-typescript-stdio')
  let g:LanguageClient_serverCommands = extend(g:LanguageClient_serverCommands, {
    \ 'javascript': ['javascript-typescript-stdio'],
    \ 'typescript': ['javascript-typescript-stdio'],
    \ 'javascript.jsx': ['javascript-typescript-stdio'],
    \ 'typescript.tsx': ['javascript-typescript-stdio'],
    \ })
endif

" pub global activate dart_language_server
if executable('dart_language_server')
  let g:LanguageClient_serverCommands = extend(g:LanguageClient_serverCommands, {
    \ 'dart': ['dart_language_server'],
    \ })
endif
```

**Comment by sandangel** [comment]
thank you. I will try it when I have time and reply with some feedback soon. ^^

**Comment by terkelg** [comment]
@sandangel did you figure out how to get this working with a language server like https://github.com/autozimu/LanguageClient-neovim?

**Comment by lifepillar** [comment]
@terkelg First of all, does your configuration work correctly when you complete with CTRL-X CTRL-O (i.e., without MUcomplete)? If that is the case, but you have some issue when MUcomplete enters the stage, I would be grateful if you could provide your feedback at issue https://github.com/lifepillar/vim-mucomplete/issues/139 or by opening a new issue.

**Comment by lifepillar** [comment]
@terkelg I have added a paragraph about compatibility with LanguageClient-neovim in the documentation. Let me know whether it helps you fix your issue.

**Comment by lifepillar** [comment]
@sandangel Sorry for coming back so late. I have now realised that my comments were more about the Reddit post you've linked to than about your code.

As for the specific languages you are using (JavaScript and TypeScript in particular): you said you were able to set up LanguageClient. Does omni-completion without MUcomplete (i.e., triggered by CTRL-X CTRL-O) work for you? If not, is`omnifunc` set to `LanguageClient#complete`?

I have tried [javascript-typescript-langserver](https://github.com/sourcegraph/javascript-typescript-langserver): I can get omni-completion to work with JavaScript (but it's slow as hell and… launching five node processes, really?!), but not with TypeScript (Vim gets stuck). I haven't tried too hard, but I think that your issue has to do with the configuration of javascript-typescript-langserver and LanguageClient. Once omni-completion is working correctly, MUcomplete should work just fine, too (but see the section on LanguageClient in MUcomplete's documentation).
